src/test.js

```diff
@@ -164,6 +165,15 @@ Test.prototype = {
 		function runTest( test ) {
 			promise = test.callback.call( test.testEnvironment, test.assert );
 			test.resolvePromise( promise );
+
+			// If the test has a "lock" on it, but the timeout is 0, then we push a
+			// failure as the test should be synchronous.
+			if ( test.timeout === 0 && test.semaphore !== 0 ) {
```

Comment:
I keep staring at `test.semaphore !== 0` here, and below at `expect( 1 )` in the failing test cases. I can't quite put my finger on it... Are they needed? What are they for in this case?